protocol/message.go

```diff
@@ -67,154 +112,76 @@ func (variant MessageVariant) String() string {
 	}
 }
 
-type MessageHash [32]byte
-
-func (hash MessageHash) String() string {
-	return base64.StdEncoding.EncodeToString(hash[:])
+// ValidateMessageVariant checks if the given variant is supported.
+func ValidateMessageVariant(variant MessageVariant) error {
+	switch variant {
+	case Ping, Pong, Cast, Multicast, Broadcast:
+		return nil
+	default:
+		return NewErrMessageVariantIsNotSupported(variant)
+	}
 }
 
+// MessageBody contains the content of the message.
 type MessageBody []byte
 
+// String implements the `Stringer` interface.
 func (body MessageBody) String() string {
-	return base64.StdEncoding.EncodeToString(body)
+	return base64.RawStdEncoding.EncodeToString(body)
 }
 
+// Message is the object users communicating in the network.
 type Message struct {
 	Length  MessageLength
 	Version MessageVersion
 	Variant MessageVariant
+	GroupID PeerGroupID
 	Body    MessageBody
 }
 
-func NewMessage(version MessageVersion, variant MessageVariant, body MessageBody) Message {
-	switch version {
-	case V1:
-	default:
-		panic(newErrMessageVersionIsNotSupported(version))
+// NewMessage returns a new message with given version, variant and body.
+func NewMessage(version MessageVersion, variant MessageVariant, groupID PeerGroupID, body MessageBody) Message {
+	if err := ValidateMessageVersion(version); err != nil {
+		panic(err)
 	}
-	switch variant {
-	case Ping, Pong, Cast, Multicast, Broadcast:
-	default:
-		panic(NewErrMessageVariantIsNotSupported(variant))
+	if err := ValidateMessageVariant(variant); err != nil {
+		panic(err)
+	}
+	if err := ValidatePeerGroupID(groupID, variant); err != nil {
+		panic(err)
 	}
 	if body == nil {
 		body = make(MessageBody, 0)
 	}
 
+	length := 8
+	if variant == Broadcast || variant == Multicast {
+		length = 40
+	}
```

Comment:
A comment explaining these magic lengths would be good. Again, should we be checking versions too (before assuming that group IDs are prefixed to the body)?